mlos_bench/mlos_bench/storage/base_storage.py

```diff
@@ -0,0 +1,231 @@
+#
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+#
+"""
+Base interface for saving and restoring the benchmark data.
+"""
+
+import logging
+from abc import ABCMeta, abstractmethod
+from typing import Optional, Union, List, Tuple, Dict, Any
+
+from mlos_bench.environment import Status
+from mlos_bench.service import Service
+from mlos_bench.tunables import TunableGroups
+from mlos_bench.util import get_git_info
+
+_LOG = logging.getLogger(__name__)
+
+# pylint: disable=too-few-public-methods,too-many-arguments
+
+
+class Storage(metaclass=ABCMeta):
+    """
+    An abstract interface between the benchmarking framework
+    and storage systems (e.g., SQLite or MLFLow).
+    """
+
+    def __init__(self, tunables: TunableGroups, service: Optional[Service], config: dict):
+        """
+        Create a new storage object.
+
+        Parameters
+        ----------
+        tunables : TunableGroups
+            Tunable parameters of the environment. We need them to validate the
+            configurations of merged-in experiments and restored/pending trials.
+        config : dict
+            Free-format key/value pairs of configuration parameters.
+        """
+        _LOG.debug("Storage config: %s", config)
+        self._tunables = tunables.copy()
+        self._service = service
+        self._config = config.copy()
+
+    @abstractmethod
+    def experiment(self, exp_id: str, trial_id: int, description: str, opt_target: str):
+        """
+        Create a new experiment in the storage.
+
+        Parameters
+        ----------
+        exp_id : str
+            Unique identifier of the experiment.
+        trial_id : int
+            Starting number of the trial.
+        description : str
+            Human-readable description of the experiment.
+        opt_target : str
+            Name of metric we're optimizing for.
+
+        Returns
+        -------
+        experiment : Storage.Experiment
+            An object that allows to update the storage with
+            the results of the experiment and related data.
+        """
+
+    class Experiment(metaclass=ABCMeta):
+        """
+        Base interface for storing the results of the experiment.
+        This class is instantiated in the `Storage.experiment()` method.
+        """
+
+        def __init__(self, tunables: TunableGroups, experiment_id: str,
+                     trial_id: int, description: str, opt_target: str):
+            self._tunables = tunables  # No need to copy, it's immutable
+            self._experiment_id = experiment_id
+            self._trial_id = trial_id
+            self._description = description
+            self._opt_target = opt_target
+            (self._git_repo, self._git_commit) = get_git_info()
+
+        def __enter__(self):
+            """
+            Enter the context of the experiment.
```

Comment:
I don't feel like these are the most self explanatory without really knowing a ton of python guts.
Just for readability, could we have other functions that subclasses must implement:

```python
@abstractmethod
def _setup(self):
    ...

@abstractmethod
def _cleanup(self):
    ...
```

And then have `__enter__` and `__exit__` call those?

Reply:
Alternatively, add something to the docstrings that makes it very clear that the `__enter__` `__exit__` functions serve those same purposes and that `Experiments` are only ever expected to be used within a `with ... as:` context.

Reply:
Good idea. I've added `_setup`/`_teardown` methods to the storage. (I don't decorate them as `@abstractmethod` because some implementations might not need to override them)